src/plan.rs

```diff
@@ -226,7 +191,8 @@ impl Recipe {
 /// safe to drop the planner after creating Fft instances.
 pub struct FftPlannerScalar<T: FftNum> {
     algorithm_cache: FftCache<T>,
-    recipe_cache: HashMap<usize, Rc<Recipe>>,
+    recipe_cache: HashMap<usize, RecipeId>,
```

Comment:
Take a look at the `FftCache` struct. I used to have just a HashMap for Fft instances - when it got a little more complicated (ie a forward cache and inverse cache), i wrapped it in a struct that hid away the implementation details of the cache. I recommend giving the recipe cache the same treatment.

Given that you can never delete recipes once they're added, I think the HashMap may be unnecessary - instead there could be a RecipeCache struct that hands out RecipeId instances which are wrappers around direct indexes into the recipe arena.

Reply:
Oh, now that I think about it more I see - the keys of the hashmap are still recipe sizes.

Maybe a RecipeId could be both recipe size and recipe index?

Reply:
The question here is: Do we ever construct two different Recipe for a given size ? I did not analyze this in depth and I took thus the conservative hypothesis of an affirmative answer.
If the answer is negative, ie, there is only ever a single Recipe build for each size, then the solution is easy: use a hashmap that maps size to Recipe, and you do not need the recipe_arena. (And the RecipeId can be the size.) I can change the code to do that.
If the answer is affirmative, since there is a one-to-one mapping between RecipeId and Recipe, we cannot take the size as the RecipeId. I believe there is no way to simplify much the code in this case.
Can you tell me in what situation we are ?

Reply:
I don't think we want to have more than one recipe for each size. My thinking is that if a a given recipe is the fastest solution for a length, then it doesn't matter if it's used directly or as the inner recipe of some other recipe, it will always be the best choice for that length.
Also having more than one recipe for a length in the cache means we would need some method for selecting which one to use. You can't just pick from the cache then.

Reply:
I was a bit confused by the fact that there is a direct construction of a Radix4 in design_prime instead of calling design_fft_with_factors, but this seems to be equivalent. Indeed then, we can take the more simple solution.